src/index.ts

```diff
@@ -135,16 +166,48 @@ class Compiler {
     }
   }
 
-  private createTsType(rule: JSONSchema.Schema): TsType.TsType {
+  private createTsType (rule: JSONSchema.Schema, propName?: string, isTop: boolean = false, isReference: boolean = false): TsType.TsTypeBase {
     switch (this.getRuleType(rule)) {
       case RuleType.AnonymousSchema:
       case RuleType.NamedSchema:
         return this.toTsDeclaration(rule)
       case RuleType.Enum:
-        return new TsType.Union(uniqBy(
-          rule.enum!.map(_ => this.toStringLiteral(_))
-          , _ => _.toType(this.settings))
-        )
+        // we honor the schema's "type" on the enum.  if string, generate a union.
+        // if int, require the tsEnumNames
+        let enumType = this.validateEnumMembers(rule)
+
+        switch (enumType){
+          case EnumType.Integer:
+            var enumValues = zip(rule.tsEnumNames || [],
+                // If we try to create a literal from an object, bad stuff can happen... so we have to toString it
+                rule.enum!.map(_ => new TsType.Literal(_).toType(this.settings).toString()))
+              .map(_ => new TsType.EnumValue(_))
```

Comment:
That validation occurs, yes, and also we validate that rule.enum is truthy before deciding that RuleType is RuleType.Enum, at line 64.

````
if (rule.enum) {
      return RuleType.Enum
    }
````